contracts/infinite-pool/RewardsPoolBaseInfinite.sol

```diff
@@ -0,0 +1,566 @@
+    }
+
+    function _applyExtension(uint256 _startTimestamp, uint256 _endTimestamp) internal {
+        uint256 rewardPerSecondLength = rewardPerSecond.length;
+        uint256 rewardsTokensLength = rewardsTokens.length;
+        uint256[] memory _rewardPerSecond = new uint256[](rewardsTokensLength);
+
+        for (uint256 i = 0; i < rewardPerSecondLength; i++) {
+            uint256 spentRewards = calculateRewardsAmount(startTimestamp, endTimestamp, rewardPerSecond[i]);
+            totalSpentRewards[i] = totalSpentRewards[i] + spentRewards;
+        }
+
+        for (uint256 i = 0; i < rewardsTokensLength; i++) {
+            uint256 balance = IERC20(rewardsTokens[i]).balanceOf(address(this));
+            // require(balance > 0, 'RewardsPoolBaseInfinite: no rewards for this token');
+
+            // we need to cut off 1% or 5% whatever the business decides
+            // IERC20(rewardsTokens[i]).transferFrom(address(this), feeRecipient, (balance * CUT_FEE) / MAX_FEE);
+
+            _rewardPerSecond[i] = (balance / PRECISION) / (_endTimestamp - _startTimestamp); // calculate the rewards per second
+
+            uint256 rewardsAmount = calculateRewardsAmount(_startTimestamp, _endTimestamp, rewardPerSecond[i]);
+
+            require(balance >= rewardsAmount, 'RewardsPoolBaseInfinite: not enough rewards');
+        }
+
+        previousCampaigns.push(Campaign(startTimestamp, endTimestamp, rewardPerSecond));
+
+        rewardPerSecond = _rewardPerSecond;
+        startTimestamp = _startTimestamp;
+        endTimestamp = _endTimestamp;
+        lastRewardTimestamp = _startTimestamp;
+
+        emit Extended(_startTimestamp, _endTimestamp, rewardPerSecond);
+    }
+
+    /**
+     *@dev Calculates the available amount of reward tokens that are not locked
+     *@param _rewardTokenIndex the index of the reward token to check
+     */
+    function getAvailableBalance(uint256 _rewardTokenIndex) public view returns (uint256) {
+        address rewardToken = rewardsTokens[_rewardTokenIndex];
+        uint256 balance = IERC20(rewardToken).balanceOf(address(this));
+
+        if (startTimestamp == 0) {
+            return balance;
+        }
+
+        uint256 spentRewards = calculateRewardsAmount(startTimestamp, endTimestamp, rewardPerSecond[_rewardTokenIndex]);
+
+        uint256 availableBalance = balance -
+            (totalSpentRewards[_rewardTokenIndex] + spentRewards - totalClaimed[_rewardTokenIndex]);
+
+        if (rewardToken == address(stakingToken)) {
+            availableBalance = availableBalance - totalStaked;
+        }
+
+        return availableBalance;
+    }
+
+    /** @dev Withdraw tokens other than the staking and reward token, for example rewards from liquidity mining
+     * @param _recipient The address to whom the rewards will be transferred
+     * @param _token The address of the rewards contract
+     */
+    function withdrawTokens(address _recipient, address _token) external onlyOwner {
```

Comment:
If users sends accidentially staking tokens to the contract, they will be stuck indefinitely if they are not part of the reward tokens

Reply:
That's the behaviour we want, if you have any idea to improve this, let me know
For now i'll close the conversation @Mr-ONADASKY